diff --git a/libsignal-service-dotnet/SignalServiceAccountManager.cs b/libsignal-service-dotnet/SignalServiceAccountManager.cs
index 191aca9..8c6049a 100644
--- a/libsignal-service-dotnet/SignalServiceAccountManager.cs
+++ b/libsignal-service-dotnet/SignalServiceAccountManager.cs
@@ -42,10 +42,10 @@ namespace libsignalservice
         /// <param name="deviceId">A Signal Service device id</param>
         /// <param name="userAgent">A string which identifies the client software</param>
         public SignalServiceAccountManager(SignalServiceConfiguration configuration,
-                                        string user, string password, int deviceId, string userAgent)
+                                        string user, string password, int deviceId, string userAgent, string number)
         {
             PushServiceSocket = new PushServiceSocket(configuration, new StaticCredentialsProvider(user, password, null, deviceId), userAgent);
-            User = user;
+            User = number;
             UserAgent = userAgent;
         }
 
